### ISSUE TITLE
Ability of Telling if User Changed Buffer when Editing History Items

### ISSUE THREAD

**Issue by mitnk** [murarth/linefeed#23]
Say we have following 3 history items:
```
$ echo 123
$ echo 222
$ echo 135
```

Let's fetch back `echo 123` this way:
```
$ echo<UP>  # and now "echo 135" shows up
# let's edit the buffer to "echo 1" from "echo 135", and then up:
$ echo 1<UP> # and now "echo 123" shows up. (wanted result, but linefeed may not able to do this)
```

I'm not sure linefeed is able to tell if the "echo 1" was edited by user, and want to "search history further with it". With `reader.buffer()` and `reader.backup_buffer()`, we cannot do it, because when we hit UP on "echo 1" as above, `buffer` is "echo 1" and `backup_buffer` is "echo" - It cannot tell whether user want to search till "echo 222" or "echo 123".

**Comment by murarth** [comment]
I hadn't realized that `history-search-forward` and `-backward` were core commands in readline that were missing in linefeed. So, I've pushed a commit that implements these commands in linefeed. It should take care of this issue and I hope I've covered the edge cases to make its behavior compatible with the equivalent readline commands.

I've also added a new method, `Reader::last_command_category`, which returns the field that is used internally to track when commands are used in succession. This may help in future situations such as this when implementing custom commands. I'm not sure that it will be totally useful in implementing custom commands, however, as the `Category` enum has only an `Other` variant which doesn't express much. But I think I'll hold off on making modifications to `Category` until an issue like this comes up again.

**Comment by mitnk** [comment]
Wow! I used to implement `history-search-forward` myself with methods of `Reader`. Now it seems I can remove them all, and linefeed can loads `inputrc` and functions properly on `history-search-forward`. Great job!

I'll test a bit more later. Thanks @murarth!

**Comment by mitnk** [comment]
Command `history-search-forward` works as expected. Since I don't have a chance to add a custom command yet, so no comments here on `Reader::last_command_category`.

Could you also release a new version on crates? Thanks again!

**Comment by mitnk** [comment]
ping on crates release

**Comment by murarth** [comment]
Just published 0.4.0.

**Comment by mitnk** [comment]
Thank you!
